qiskit/synthesis/multi_controlled/mcx_synthesis.py

```diff
@@ -300,6 +301,308 @@ def synth_mcx_noaux_v24(num_ctrl_qubits: int) -> QuantumCircuit:
     return qc
 
 
+def linear_depth_ladder_ops(qreg: List[int]) -> Tuple[QuantumCircuit, List[int]]:
+    r"""
+    Helper function to create linear-depth ladder operations used in Khattar and Gidney's MCX synthesis.
+    In particular, this implements Step-1 and Step-2 on Fig. 3 of [1] except for the first and last
+    CCX gates.
+
+    Args:
+        qreg: List of qubit indices to apply the ladder operations on. qreg[0] is assumed to be ancilla.
+
+    Returns:
+        QuantumCircuit: Linear-depth ladder circuit.
+        int: Index of control qubit to apply the final CCX gate.
+
+    References:
+        1. Khattar and Gidney, Rise of conditionally clean ancillae for optimizing quantum circuits
+        `arxiv:2407.17966 https://arxiv.org/abs/2407.17966`_
+    """
+
+    n = len(qreg)
+    assert n > 3, "n = n_ctrls + 1 => n_ctrls >= 3 to use MCX ladder. Otherwise, use CCX"
```

Comment:
We don't use assertions in code to be released, you could instead raise a `ValueError` (or handle the cases with 3 qubits or less, if it could occur)